sound/soc/sof/debug.c

```diff
@@ -17,6 +17,234 @@
+				    unsigned long ipc_duration_ms,
+				    unsigned long ipc_count)
+{
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	u64 min_response_time = U64_MAX;
+	ktime_t start, end, test_end;
+	u64 avg_response_time = 0;
+	u64 max_response_time = 0;
+	u64 ipc_response_time;
+	char  *output;
+	int len2, len3;
+	int len1 = 0;
+	int i = 0;
+	int size;
+	int ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST_MSG | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * NSEC_PER_MSEC;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &reply, sizeof(reply));
+		end = ktime_get();
+
+		if (ret < 0)
+			break;
+
+		/* compute min and max response times */
+		ipc_response_time = ktime_to_ns(ktime_sub(end, start));
+		min_response_time = min(min_response_time, ipc_response_time);
+		max_response_time = max(max_response_time, ipc_response_time);
+
+		/* sum up response times */
+		avg_response_time += ipc_response_time;
+		i++;
+
+		/* test complete? */
+		if (flood_duration_test) {
+			if (ktime_to_ns(end) >= test_end)
+				break;
+		} else {
+			if (i == ipc_count)
+				break;
+		}
+	}
+
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: ipc flood test failed at %d iterations\n", i);
+
+	/* compute average response time */
+	do_div(avg_response_time, i);
+
+	if (flood_duration_test) {
+		dev_dbg(sdev->dev, "IPC Flood test duration: %lums\n",
+			ipc_duration_ms);
+		len1 = snprintf(NULL, 0,
+				"IPC Flood test duration: %lums\n",
+				ipc_duration_ms);
+	}
+
+	/* clear previous test output */
+	memset(dfse->cache_buf, 0, IPC_FLOOD_TEST_RESULT_LEN);
+
+	dev_dbg(sdev->dev,
+		"IPC Flood count: %d, Avg response time: %lluns\n",
+		i, avg_response_time);
+	dev_dbg(sdev->dev, "Max response time: %lluns\n",
+		max_response_time);
+	dev_dbg(sdev->dev, "Min response time: %lluns\n",
+		min_response_time);
+
+	/* format test output */
+	len2 = snprintf(NULL, 0,
+			"IPC Flood count: %d\nAvg response time: %lluns\n",
+			i, avg_response_time);
+
+	len3 = snprintf(NULL, 0,
+			"Max response time: %lluns\nMin response time: %lluns\n",
+			max_response_time, min_response_time);
+
+	size = len1 + len2 + len3 + 1;
+
+	output = kzalloc(size, GFP_KERNEL);
+	if (!output)
+		return -ENOMEM;
+
+	if (flood_duration_test)
+		sprintf(output, "IPC Flood test duration: %lums\n",
+			ipc_duration_ms);
+
+	sprintf(output + len1,
+		"IPC Flood count: %d\nAvg response time: %lluns\n",
+		i, avg_response_time);
+
+	sprintf(output + len1 + len2,
+		"Max response time: %lluns\nMin response time: %lluns\n",
+		max_response_time, min_response_time);
+
+	/* find the minimum. min() is not used to avoid warnings */
+	if (size > IPC_FLOOD_TEST_RESULT_LEN)
+		size = IPC_FLOOD_TEST_RESULT_LEN;
+
+	/* save current test output */
+	memcpy(dfse->cache_buf, output, size);
```

Comment:
I am missing something, what is this cache_buf? is this related to the suspend stuff?

Reply:
@plbossart cache_buf is used for the IO type debugfs but unused for buf type debugfs. So I am just using it here to store the test result. Maybe I should add a comment to clarify

Reply:
@plbossart fixed now

Reply:
@ranj063 hm, sorry, I didn't follow or re-read the complete discussion, what's preventing you from doing `snprintf(dfse->cache_buf, IPC_FLOOD_TEST_RESULT_LEN, ...)` directly?